structures.py

```diff
@@ -1548,6 +1549,105 @@ def _repr(self):
         return 'Computed({!r})'.format(self.value)
 
 
+class BitFields(Construct):
+    r"""
+    Build and parse named bit-wise fields. Values are always built from
+    unsigned big-byteorder integers and parsed as unsigned
+    big-byteorder integers.
+
+        >>> b = BitFields('version:4, header_length:4')
+        >>> b
+        BitFields('version:4, header_length:4')
+        >>> b.build({'version': 4, 'header_length': 0})
+        b'\x80'
+        >>> b.parse(b'\x00') == {'version': 0, 'header_length': 0}
+        True
+        >>> b.sizeof()
+        1
+
+    Fields can span over 8 bits, and the whole construct is byte-aligned
+    (i.e. it will read as many bytes as needed and ignore trailing bits that
+    do not have definitions in the spec):
+
+        >>> b = BitFields('foo:12,bar:5')
+        >>> b.sizeof()
+        3
+        >>> b.build({'foo': 4095, 'bar': 31})
+        b'\xff\xff\x80'
+        >>> b.parse(b'\x09\x11\x00') == {'foo': 145, 'bar': 2}
+        True
+
+    You can omit any field during building, in that case 0 will be built.
+    That allows to emulate padding that comes before actual data:
+
+        >>> b = BitFields('padding:7, flag:1')
+        >>> b.parse(b'\x01') == {'padding': 0, 'flag': 1}
+        True
+        >>> b.build({'flag': 0})
+        b'\x00'
+
+    If you try to build from integer that can't be packed into the specified
+    amount of bits, a BuildingError will be raised:
+
+        >>> b.build({'flag': 10})
+        Traceback (most recent call last):
+        ...
+        structures.BuildingError: cannot pack 10 into 1 bit
+
+    :param spec: Fields definition, a comma separated list of
+    name:length-in-bits pairs. Spaces between commas are allowed.
+
+    :param embedded: If True, this construct will be embedded into the
+    enclosed struct.
+
+    """
+
+    def __init__(self, spec, embedded=False):
+        super().__init__()
+        self.spec = spec
+        self._embedded = embedded
+        self.fields = OrderedDict()
+        for field in map(str.strip, spec.split(',')):
+            name, length = field.split(':')
+            self.fields[name] = int(length)
```

Comment:
I guess, you expect here only positive integers and field spec like `foo:-1` is invalid.